### PR TITLE
Feat/load completion improvements

### PR DESCRIPTION
**Pull request by scop**
Intended to be merged after #923.

This started from looking into https://github.com/scop/bash-completion/pull/917#discussion_r1161092202 and ballooned in scope somewhat. The gist of this is that we now provide the full command name (even when not invoked with one) to the completions when sourced, to facilitate generating from the correct executable, and for improved handling of commands invoked with relative paths.

The commits are kind of intertwined, but I hope looking into them individually for details provides enough context if the end result is hard to follow.

### CODE REVIEW

File: bash_completion [hunk 1]
```diff
@@ -2523,68 +2538,71 @@ __load_completion()
...
+                        # If invoked without path, that one should be set, too
+                        # ...but let's not overwrite an existing one, if any
+                        [[ $origcmd != */* ]] &&
+                            ! complete -p "$origcmd" &>/dev/null &&
+                            extspecs+=("$origcmd")
+                        ((${#extspecs[*]} != 0)) && $compspec "${extspecs[@]}"
+                        return 0
+                    fi
+                    # If not, see if we got one for $cmdname
+                    if compspec=$(complete -p "$cmdname" 2>/dev/null); then
```
Comment (akinomyoga) [inline bash_completion:2592]:
```suggestion
                    if [[ $cmdname != "$cmd" ]] && compspec=$(complete -p "$cmdname" 2>/dev/null); then
```

File: bash_completion [hunk 2]
```diff
@@ -2496,9 +2496,24 @@ complete -F _minimal ''
...
+        # If we already have a completion for the "real" command, use it
+        $(complete -p "$cmd" 2>/dev/null || echo false) "\\$cmd" && return 0
+        backslash=\\
+    fi
+
+    # Try to resolve real path to $cmd, and make it absolute
+    local ret realcmd="" origcmd=$cmd
+    if _comp_realcommand "$cmd"; then
+        realcmd=$ret
+        cmd=${realcmd%/*}/$cmdname # basename could be an alias
```
Comment (akinomyoga) [inline bash_completion:2515]: I think I fail to understand this line. I naively thought this should be `cmd=$realcmd` because the command name can be different in the real location, such as `~/bin/b -> /usr/bin/bash`. There seems to be a reason related to an alias as in the code comment, but I'm not sure what it exactly is now.

**Review by akinomyoga — COMMENTED** [review]
Thank you for working on this issue! This is great.

**Comment by akinomyoga** [comment]
I think we should also fix

```diff
diff --git a/completions/_nox b/completions/_nox
index ac387ef7..8e98002a 100644
--- a/completions/_nox
+++ b/completions/_nox
@@ -9,6 +9,8 @@ eval -- "$(
         register-python-argcomplete3 --shell bash "$1" 2>/dev/null
 )"

-complete -p "$1" &>/dev/null
+{
+    complete -p "$1" || complete -p "${1##*/}"
+} &>/dev/null

 # ex: filetype=sh
```

File: bash_completion (hunk 1 above)
Comment (scop) [reply bash_completion:2592]: This was my intuition too when writing it, but because we test
```bash
if compspec=$(complete -p "$cmd" 2>/dev/null); then
```
...above and return if that holds, we have already returned before reaching here if the two are equal.

Comment (akinomyoga) [reply bash_completion:2592]: I think the behavior doesn't change but we may reduce the fork. When the two are equal *and* `complete -p cmd` fails, we are going to run the subshell `$(complete -p cmd)` twice with the current version.

File: bash_completion (hunk 2 above)
Comment (scop) [reply bash_completion]: "Alias" is actually not a good word to describe this, but I think you got the intent right: the basename of the resolved command can be different from the one we were invoked with.

Installing our completion for the "real" resolved name would be problematic for example in cases where the target command has different behavior depending on the basename it is invoked as. `busybox` is one example of this.

Elaborated on the commentary in 3f479c173b9bd2619962a47b120e48acd54d6c0b

Comment (scop) [reply bash_completion]: Hmm, good point. I suppose we should actually resolve absolute path to the command without doing any symlink resolution, essentially what `_comp_realcommand` now does, but without doing any of the realpath/readlink things?

File: bash_completion (hunk 1 above)
Comment (scop) [reply bash_completion]: Indeed, applied, thanks!

File: bash_completion (hunk 2 above)
Comment (akinomyoga) [reply bash_completion]: Ah, yes. I think that is the solution.

The original reason that we wanted to resolve the symbolic links is to pick up the corresponding `share/bash-completion/completions`, so the symlink-resolved `realcmd` is still needed. So the conversion of `cmd` to an absolute path should be done independently of `realcmd`.

Comment (scop) [reply bash_completion]: 28b587a47058395fdacb7eafe866685af1e74e3c and e4ede514249c4250dd43139f54fa3c46554a1dbc are a quick attempt at fixing this, more or less untested.

File: bash_completion [hunk 3]
```diff
@@ -1659,6 +1659,25 @@ _fstypes()
     [[ $fss ]] && COMPREPLY+=($(compgen -W "$fss" -- "$cur"))
 }
 
+# Get absolute path to a file, with rudimentary canonicalization.
+# No symlink resolution or existence checks are done;
+# see `_comp_realcommand` for those.
+# @param $1 The file
+# @var[out] ret The path
+_comp_abspath()
```
Comment (scop) [inline bash_completion:1667]: This would deserve some unit tests.

File: bash_completion (hunk 2 above)
Comment (scop) [reply bash_completion]: Need to call it a day now, revisiting later. But do feel free to push changes you see fit here in the meantime if you like.

File: bash_completion [hunk 4]
```diff
@@ -2523,68 +2548,71 @@ __load_completion()
         dirs+=(./completions)
     fi
 
-    # 3) From bin directories extracted from $(realpath "$cmd") and PATH
-    local ret
-    _comp_realcommand "$1" && paths=("${ret%/*}") || paths=()
+    # 3) From bin directories extracted from path to command, and $PATH
+    [[ $cmd == */* ]] && paths=("${cmd%/*}") || paths=()
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply bash_completion:2552]: I believe we still need to use the symlink-resolved `readlcmd` here. We wanted to find `share/bash-completion/completions` of the actual command even if it is called from a symlink. This is independent of how we resolve `cmd`.

File: bash_completion (hunk 2 above)
Comment (akinomyoga) [reply bash_completion]: Maybe I have confused you with my previous reply, but I actually agreed to switch to the absolute path without resolving the symbolic links. I wanted to note that the symlink-resolved `realcmd` is still used later for another purpose.

File: bash_completion (hunk 4 above)
[the remark this answers is not in the document]
Comment (akinomyoga) [reply bash_completion:2552]: > Unsure if this check should be `== /*` instead of `== */*`. But I suppose it doesn't make much of a difference.

A relative path containing a slash can happen here only when the executable file does not exist. The question is whether we should search the corresponding completion file in the relative-path directory. I'm not sure it's worth it but added 301da7f6.

> I believe we still need to use the symlink-resolved `readlcmd` here. We wanted to find `share/bash-completion/completions` of the actual command even if it is called from a symlink. This is independent of how we resolve `cmd`. (by @akinomyoga)

I have pushed a fix for the real path bbe1ae88.

File: bash_completion [hunk 5]
```diff
@@ -2496,9 +2506,24 @@ complete -F _minimal ''
...
+    if [[ $cmd == \\* ]]; then
+        cmd=${cmd:1}
+        # If we already have a completion for the "real" command, use it
+        $(complete -p "$cmd" 2>/dev/null || echo false) "\\$cmd" && return 0
+        backslash=\\
+    fi
+
+    # Resolve absolute path to $cmd
+    local ret pathcmd origcmd=$cmd
+    if pathcmd=$(type -P "$cmd"); then
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply bash_completion:2523]: I think either work. I don't have a particular preference here.

File: bash_completion (hunk 4 above)
[the remark this answers is not in the document]
Comment (scop) [reply bash_completion:2552]: Hm, I'm unable to locate the mentioned commits in the repo. Were they pushed to a fork somewhere? Feel free to push them in the branch here.

[the remark this answers is not in the document]
Comment (akinomyoga) [reply bash_completion]: Ah, sorry. It's just my mistake. I intended to push them into the PR branch but have actually pushed them into the branch of the same name in my repository. I now pushed them.

File: bash_completion (hunk 3 above)
Comment (scop) [reply bash_completion:1667]: Added in 81132e79a9902783c019a941d9568f19a8c34de2.

File: bash_completion [hunk 6]
```diff
@@ -2523,68 +2548,75 @@ __load_completion()
...
     fi
 
-    # 3) From bin directories extracted from $(realpath "$cmd") and PATH
+    # 3) From bin directories extracted from the real path to the command, the
+    # specified path to the command, and $PATH
     local ret
-    _comp_realcommand "$1" && paths=("${ret%/*}") || paths=()
+    paths=()
+    _comp_realcommand "$cmd" && paths+=("${ret%/*}")
+    [[ $cmd == /* ]] && paths+=("${cmd%/*}")
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply bash_completion:2556]: Hmm, I agree.

[the remark this answers is not in the document]
Comment (akinomyoga) [reply bash_completion]: I modified the commit bbe1ae88 (diff [`bbe1ae88..61982cbc`](https://github.com/scop/bash-completion/compare/bbe1ae88..61982cbc)) and force-pushed.

**Review by akinomyoga — APPROVED** [review]
